File: 637.二叉树的层平均值.cpp
/*
 * @lc app=leetcode.cn id=637 lang=cpp
 *
 * [637] 二叉树的层平均值
 */
#include <bits/stdc++.h>
#include <ranges>
using namespace std;
struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<double> averageOfLevels(TreeNode* root) {
        vector<double> res;
        int front = 0, rear = 0, end = 0;
        TreeNode* q[10005];
        if (root) q[rear ++] = root, end ++;
        double mean = 0;
        int cnt = 0;
        while (front < rear)
        {
            cnt ++;
            TreeNode* tmp = q[front ++];
            mean += tmp->val;
            if (tmp->left) q[rear ++] = tmp->left;
            if (tmp->right) q[rear ++] = tmp->right;
            if (front == end)
            {
                res.push_back(mean / cnt);
                cnt = 0, mean = 0;
                end = rear;
            }
        }
        return res;
    }
};
// @lc code=end

File: 面经150/面经150p1/10.二叉树层次遍历/637.二叉树的层平均值.py
#
# @lc app=leetcode.cn id=637 lang=python3
#
# [637] 二叉树的层平均值
#
from mytools import *
# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def averageOfLevels(self, root: Optional[TreeNode]) -> List[float]:
        q = deque([root])
        res = []
        while q:
            n = len(q)
            s = 0
            for _ in range(n):
                tmp = q.popleft()
                s += tmp.val
                if tmp.left: q.append(tmp.left)
                if tmp.right: q.append(tmp.right)
            res.append(s / n)
        return res
# @lc code=end

